gptme/tools/time_awareness.py

```diff
@@ -0,0 +1,129 @@
+"""
+Time awareness tool.
+
+Provides time feedback during conversations to help the assistant manage
+long-running sessions effectively.
+
+Shows time elapsed messages at: 1min, 5min, 10min, 15min, 20min, then every 10min.
+"""
+
+import logging
+from collections.abc import Generator
+from datetime import datetime
+from pathlib import Path
+
+from ..hooks import HookType
+from ..logmanager import Log
+from ..message import Message
+from .base import ToolSpec
+
+logger = logging.getLogger(__name__)
+
+# Track conversation start times per workspace
+_conversation_start_times: dict[str, datetime] = {}
+
+# Track which time milestones have been shown per workspace
+_shown_milestones: dict[str, set[int]] = {}
+
+
+def add_time_message(
+    log: Log, workspace: Path | None, manager=None
+) -> Generator[Message, None, None]:
+    """Add time elapsed message after message processing.
+
+    Shows messages at: 1min, 5min, 10min, 15min, 20min, then every 10min.
+    """
+    try:
+        if workspace is None:
+            return
+
+        workspace_str = str(workspace)
+
+        # Initialize conversation start time if first message
+        if workspace_str not in _conversation_start_times:
+            _conversation_start_times[workspace_str] = datetime.now()
+            _shown_milestones[workspace_str] = set()
+            return
+
+        # Calculate elapsed time in minutes
+        elapsed = datetime.now() - _conversation_start_times[workspace_str]
+        elapsed_minutes = int(elapsed.total_seconds() / 60)
+
+        # Determine which milestone to show
+        milestone = _get_next_milestone(elapsed_minutes)
+
+        # Check if we should show this milestone
+        if milestone and milestone not in _shown_milestones[workspace_str]:
+            _shown_milestones[workspace_str].add(milestone)
+
+            # Format time message
+            hours = elapsed_minutes // 60
+            minutes = elapsed_minutes % 60
+
+            if hours > 0:
+                time_str = f"{hours}h {minutes}min" if minutes > 0 else f"{hours}h"
+            else:
+                time_str = f"{minutes}min"
+
+            message = Message(
+                "system",
+                f"<system_warning>Time elapsed: {time_str}</system_warning>",
+                hide=True,
+            )
```

Comment:
I think "Time elapsed" could be problematic in certain situations. Conversations can be resumed in which case the `_conversation_start_times` and `_shown_milestones` would be empty. I think we should rather print out a "The time is now HH:MM" (24h format) and possibly compute a time elapsed (or not? let the agent infer) from the timestamps in the message log (each `Message` has a timestamp).